src/main/scala/loamstream/loam/LoamGraph.scala

```diff
@@ -201,12 +203,20 @@ final case class LoamGraph(
 
     val toolOutputsNew = toolOutputs + (tool -> (outputsFor(tool) -- stores))
 
-    val storeProducersNew = storeProducers.filterNot {
-      case (store, producer) => stores(store) && producer == tool
+    //Remove only at most `stores.size` keys from `storeProducers`.  Previously, this code relied on calling
+    //`storeProducers.filterNot`, which traversed the entire storeProducers Map and ran in O(nStores) time.
+    //The current approach is faster since for non-trivial pipelines, the number of keys to remove from 
+    //storeProducers is usually *much* smaller than the total number of stores, and removing a key from a map is
+    //fast, either O(1) or O(log nStores).
+    val storeProducersKeysToRemove: Iterable[Store] = {
+      stores.map { s => s -> storeProducers.get(s) }.collect { case (s, Some(producer)) if producer == tool => s }
     }
+    
+    val storeProducersNew = storeProducers -- storeProducersKeysToRemove
 
-    val storeConsumersNew =
+    val storeConsumersNew = {
       storeConsumers ++ stores.map(store => (store, storeConsumers.getOrElse(store, Set.empty) + tool))
+    }
```

Comment:
At this point it seems like you compute the producer stores to remove/new and then do the same for consumers? I don't know the typical size of `stores`, but if large, since this is a performance area, maybe worth considering doing all those operations together (`O(n)` as opposed to `c*O(n)`)? 

Given the gains you've already gotten, it's likely not worth it, but it stood out so I thought I'd mention it.

Reply:
Profiling didn't turn this bit up, but it might rear its head eventually.  I think the biggest factor here is that the size of `stores` is usually small (think 10s at most).